{% extends "base.html" %}

{% block body %}


     <div id="cite" class="field">
        <div class="top">
            <div class="description"><h2>{% if help %}Run {{ SITE_NAME }}{% else %}{{ this|capfirst }}{% endif %}</h2></div>
            <div class="line tall"><hr /></div>
        </div>
        <div class="content">
        
        
        	{% if help %}
        	
        		<p>On the job submission page, users would see the page as below.</p>
        		<div class="helpimg"><img src="{{ STATIC_FOLDER }}images/help1bwedit.png" /></div>
        		<div class="helptxt">
            		<h4>1. Enter protein identifier</h4>
            		<p class="pul">
            		    Any protein in the <a class="click2" href="http://www.uniprot.org/" target="_blank">Uniprot knowledgebase</a> 
            		    (>50M) can be entered with the most common identifiers, gene names, and synonyms (>800M). 
            		    Example input: <span class="mono click2">OO0522</span>, 
        		                       <span class="mono click2">CCM1</span>,
        		                       <span class="mono click2">KRIT1_HUMAN</span>,
        		                       <span class="mono click2">ENSP00000391675</span>, 
        		                       <span class="mono click2">GC07M091828</span>.
            		</p>
        		    <ul><li>Warning: as only structural models for the human proteome has been precalculated, predictions for 
        		            other organisms can be expected to last multiple hours per mutation.
        		        </li>
        		    </ul>
            		<h4>2. Select mutation in protein</h4>
            		<p class="pul">
            		    Mutations can be selected by either clicking the protein figure or by using the select option next to the 
            		    protein input.
            		</p>
            		<ul><li>Warning: because ELASPIC uses structures of protein domains to predict stability/affinity effects, the 
            		        mutation has to fall within domain boundries.
        		        </li>
        		    </ul>
            		<h4>3. Submit job</h4>
            		<p>A valid e-mail address has to be entered to submit a job. An email will be sent on job start and completion 
            		with a link to the results. Check junk e-mail folder just in case the notification got delivered there. </p>
            		<h4>! Switch to batch input</h4>
            		<p>An option to input multiple mutations at once or upload a file with data is available.</p>
            		<h4>!! See previously run mutations</h4>
            		<p>Mutations already in the database are shown on the protein figure. Click each individual mutation to see their 
            		results in the same format as on the result page.</p>
        		</div>
        </div>
    </div>
    <div id="help2" class="field">
        <div class="top">
            <div class="description"><h2>{{ SITE_NAME }} prediction results</h2></div>
            <div class="line tall"><hr /></div>
        </div>
        <div class="content">
        		<p>As output, ELASPIC gives a table of selected mutation and their predicted stability effects.</p>
        		<div class="helpimg"><img src="{{ STATIC_FOLDER }}images/help2bwedit.png" /></div>
        		<div class="helptxt">
            		<h4>1. Result table</h4>
            		<p class="pul">
            		    ELASPIC distinguishes between two types of mutations: 
            		        <i class="click2">core</i> <small>(of domain)</small> and 
            		        <i class="click2">interface</i> <small>(between two proteins known to interact)</small></i>.
            		    Besides mutation type, the table has six colums with results: 
            		        <b>(1-2)</b> The sequence identity and alignment score of the template used to model the structure needed. 
            		        <b>(3)</b> The DOPE score of the model given by <a class="click2" target="_blank" href="https://salilab.org/modeller/">Modeller</a>. 
            		        <b>(4-5)</b> The Gibbs free energy of the optimized wild-type and mutated models calculated by 
            		              <a class="click2" target="_blank" href="http://foldx.crg.es/">FoldX</a>. 
            		        <b>(6)</b> The final change in domain stability (core) or binding affinity (interface) as predicted by {{ SITE_NAME }}.
            		</p>
            		<ul><li>Notice: For interface mutations two values are given for both sequence identity and alignment score.
            		        The first one given will always be for the protein entered while the second is for the interacting protein.
            		        Furthermore, in the interface multiple interactions might be affected, and multiple rows will then
            		        be shown in the result table for the same mutation.
        		        </li>
        		    </ul>
            		<h4>2. Filter results</h4>
            		<p class="pul"> 
            		    The result table can be filtered to only show the specific rows desired. Filtering can be done by protein name,
            		    status of calculations, type of mutation, and by result scores.
            		</p>
            		<ul><li>Notice: Filtering not only filters the table, but also the downloadable files.
        		        </li>
        		    </ul>
            		<h4>3. Downloadable files</h4>
            		<p>
            		    The results can be downloaded in a tab-delimited text file with either only the final predicted
            		    energy change or with every single feature used by {{ SITE_NAME}}. Furthermore, created models, alignments,
            		    and sequences can be downloaded.
            		</p>
            		<h4>4. See detailed results for mutation</h4>
            		<p>Each row can be clicked to see detailed results and models for a specific mutation.</p>
        		</div>
        		<h3>Detailed results</h3>
        		<div class="helpimg"><img src="{{ STATIC_FOLDER }}images/help3bwedit.png" /></div>
        		<div class="helptxt">
            		<h4>1. Affected interaction network</h4>
            		<p class="pul">
            		    All interactions affected by the mutation are included in the network. Each
            		    interaction can be clicked to display its detailed results. 
            		</p>
            		<ul><li>Notice: For core mutations the interaction network is not shown.
        		        </li>
        		    </ul>
            		<h4>2. Protein interaction figure</h4>
            		<p class="pul">
            		    The selected interaction is shown as a figure with all protein domains. The interacting domains
            		    are color-coded with blue for entered protein and pink for interacting protein.
            		</p>
            		<ul><li>Notice: For core mutations only the entered protein is shown.
        		        </li>
        		    </ul>
            		<h4>3. Modelled structure</h4>
            		<p>
            		    A <a class="click2" target="_blank" href="http://sourceforge.net/projects/jsmol/">Jmol plugin</a> shows both the modelled wild-type and mutant 
            		    structures. 
            		    See mouse controls <a class="click" target="_blank" href="http://wiki.jmol.org/index.php/Mouse_Manual">here</a>.
            		</p>
            		<div class="helptxt">
                		<h4>! Set quality settings</h4>
                		<p>
                		    Jmol runs by default in in Java mode. However, if Java not is installed or compatible with the browser 
                		    (as on mobile phones the plugin can be run in HTML5 mode with a single click. Furthermore, if the 
                		    structural rendering is too demanding for the browser, the quality can be reduced.
                		</p>
                		<h4>!! Set display settings</h4>
                		<p>
                		    The buttons below the plugin can change what is displayed. 
                		    To control details such as labels, colors, and schemes use the buttons on the left. 
                		    To view or hide a model or protein, or to view the structures in a larger window use the buttons on the right.
                		</p>
            		</div>
        		</div>
        
        	
        	{% endif %}


			{% if reference %}
		        <p class="pdesc">If you use the result generated by {{ SITE_NAME }}, please cite the paper:</p>
		        <p class="paper"><a style="click2" target="_blank" href="http://www.ncbi.nlm.nih.gov/pubmed/25243403">
		        	Niklas Berliner, Joan Teyra, Recep Ã‡olak, Sebastian Garcia Lopez, and Philip M. Kim.
		        	Combining structural modeling with ensemble machine learning to accurately predict 
		        	protein fold stability and binding affinity effects upon mutation.
		        	<i>PloS one</i>
		        	<b>9</b>:e107353 (2014)
		        </a></p>
            {% endif %}
            
            
            {% if contact %}
            {% block headscript %}
            	<script type="text/javascript" src="{{ STATIC_FOLDER }}js/jquery.color-2.1.2.min.js"></script>
            {% endblock %}
				<form id="cont" action="#contact" method="POST">
					<div class="cont">Name:
						<input class="fname" type="text" placeholder="Name"  />
					</div>
					<div class="cont">Email:
						<input class="fmail" type="text" placeholder="Email" />
					</div>
					<div class="cont">Subject:
						<input class="ftitle" type="text" placeholder="Subject" />
					</div>
					<div class="cont float">Your message:
						<textarea class="fmsg" placeholder="Message"></textarea>
					</div>
					<div class="error"></div>
					<div class="form-response"></div>
					<input class="submit" type="submit" value="Send" />
				</form>
            {% endif %}
            
        </div>
    </div> 


{% endblock %}
